cmd/outline-ss-server/metrics.go

```diff
@@ -49,11 +61,100 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+type activeClient struct {
+	IPKey           IPKey
+	connectionCount int
+	startTime       time.Time
+}
+
+func (c *activeClient) IsActive() bool {
+	return c.connectionCount > 0
+}
+
+type IPKey struct {
+	ip        string
+	accessKey string
+}
+
+type activeIPKeyTracker struct {
+	activeClients   map[IPKey]activeClient
+	metricsCallback func(IPKey, time.Duration)
+}
+
+// Reports time connected for all active clients, called at a regular interval.
+func (t *activeIPKeyTracker) reportAll() {
+	if len(t.activeClients) == 0 {
+		logger.Debugf("No active clients. No IPKey activity to report.")
+		return
+	}
+	for _, c := range t.activeClients {
+		t.reportDuration(c)
+	}
+}
+
+// Reports time connected for a given active client.
+func (t *activeIPKeyTracker) reportDuration(c activeClient) {
+	connDuration := since(c.startTime)
+	logger.Debugf("Reporting activity for key `%v`, duration: %v", c.IPKey.accessKey, connDuration)
+	t.metricsCallback(c.IPKey, connDuration)
+
+	// Reset the start time now that it's been reported.
+	c.startTime = time.Now()
+	t.activeClients[c.IPKey] = c
+}
+
+// Registers a new active connection for a client [net.Addr] and access key.
+func (t *activeIPKeyTracker) startConnection(addr net.Addr, accessKey string) {
+	hostname, _, _ := net.SplitHostPort(addr.String())
+	ipKey := IPKey{ip: hostname, accessKey: accessKey}
+
+	c, exists := t.activeClients[ipKey]
+	if !exists {
+		c = activeClient{ipKey, 0, time.Now()}
+	}
+	c.connectionCount++
+	t.activeClients[ipKey] = c
+}
+
+// Removes an active connection for a client [net.Addr] and access key.
+func (t *activeIPKeyTracker) stopConnection(addr net.Addr, accessKey string) {
```

Comment:
This method doesn't seem to be safe to be called from concurrent go routines. It probably needs to be fixed.

Make sure you run the tests with -race.